pkg/strategy/dca2/recover.go

```diff
@@ -60,74 +64,89 @@ func (s *Strategy) recover(ctx context.Context) error {
 }
 
 // recover state
-func recoverState(ctx context.Context, maxOrderCount int, currentRound Round, orderExecutor *bbgo.GeneralOrderExecutor) (State, error) {
+func recoverState(currentRound Round, orderExecutor *bbgo.GeneralOrderExecutor) (State, error) {
+	// no open-position orders and no take-profit orders means this is the whole new strategy
+	if len(currentRound.OpenPositionOrders) == 0 && len(currentRound.TakeProfitOrders) == 0 {
+		return WaitToOpenPosition, nil
+	}
+
+	// it should not happen
+	if len(currentRound.OpenPositionOrders) == 0 && len(currentRound.TakeProfitOrders) > 0 {
+		return None, fmt.Errorf("there is no open-position orders but there are take-profit orders. it should not happen, please check it")
+	}
+
 	activeOrderBook := orderExecutor.ActiveMakerOrders()
 	orderStore := orderExecutor.OrderStore()
 
-	// dca stop at take-profit order stage
 	if len(currentRound.TakeProfitOrders) > 0 {
-		openedOrders, cancelledOrders, filledOrders, unexpectedOrders := types.ClassifyOrdersByStatus(currentRound.TakeProfitOrders)
-
-		if len(unexpectedOrders) > 0 {
-			return None, fmt.Errorf("there is unexpected status in orders %+v", unexpectedOrders)
-		}
+		return recoverStateIfAtTakeProfitStage(currentRound.TakeProfitOrders, activeOrderBook, orderStore)
+	}
 
-		if len(filledOrders) > 0 && len(openedOrders) == 0 {
-			return WaitToOpenPosition, nil
-		}
+	return recoverStateIfAtOpenPositionStage(currentRound.OpenPositionOrders, activeOrderBook, orderStore)
+}
 
-		if len(filledOrders) == 0 && len(openedOrders) > 0 {
-			// add opened order into order store
-			for _, order := range openedOrders {
-				activeOrderBook.Add(order)
-				orderStore.Add(order)
-			}
-			return TakeProfitReady, nil
-		}
+// recoverStateIfAtTakeProfitStage will recover the state if the strategy is stopped at take-profit stage
+func recoverStateIfAtTakeProfitStage(orders []types.Order, activeOrderBook *bbgo.ActiveOrderBook, orderStore *core.OrderStore) (State, error) {
+	// because we may manually recover the strategy by cancelling the orders and placing the orders again, so we don't need to consider the cancelled orders
+	openedOrders, cancelledOrders, filledOrders, unexpectedOrders := types.ClassifyOrdersByStatus(orders)
```

Comment:
move this method ClassifyOrdersByStatus to types.OrderSlice as ClassifyByStatus ?